src/feature/flow_chart_panel/views/ControlBar.tsx

```diff
@@ -215,6 +215,7 @@ const LoadButton = () => {
 
 const ControlBar = () => {
   const { states } = useSocket();
+  const { nodes, edges } = useFlowChartGraph();
```

Comment:
Avoid calling this hook wherever possible, it makes the component rerender (with all of its children) every single time a node changes (so every time a node is dragged by a pixel in the editor). Only the buttons that need access to this state call this hook so the rerenders are isolated to them specifically, to minimize rerenders. You could just put the keyboard shortcuts in their respective buttons since they already have access to the state.

Reply:
@JeffDotPng 
So instead of calling it like this:   
```
useKeyboardShortcut("ctrl", "p", () => onRun(nodes, edges));
  useKeyboardShortcut("ctrl", "c", cancelFC);
  useKeyboardShortcut("ctrl", "s", () => saveFile(nodes, edges));
```
You are suggesting that the callback function should be related to the respective buttons?

Reply:
Just put the hook in the respective buttons, they already have to get `useFlowChartGraph` to get the necessary state, avoiding `ControlBar` constantly rerendering.